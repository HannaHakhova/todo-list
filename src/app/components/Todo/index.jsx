import React, {Component} from 'react';
import {Checkbox, List, Button, Label, Icon} from 'semantic-ui-react';
import {shape, func, string, instanceOf, number, bool, arrayOf} from 'prop-types';
import {getLabelDate} from 'utils';
import TodoTagList from 'components/TodoTagList';

class Todo extends Component {
    onDrop = e => {
        const data = e.dataTransfer.getData('text/plain');
        this.props.addTodoTag(data, this.props.todo.id);
    }
    allowDrop = ev => {
        ev.preventDefault();
    }

    render() {
        const tagArray = this.props.todo.tags;
        return (
            <div onDragOver={this.allowDrop} onDrop={this.onDrop}>
                <div>
                    <List.Item floated='left'>
                        <Checkbox
                            checked={this.props.todo.completed}
                            onChange={() => this.props.handleComplete(this.props.todo.id)}
                            label={this.props.todo.title} />
                        <Label as='a' size='tiny' onClick={() => this.props.sortByDateAndCompleted()}>
                            <Icon name='calendar' />
                            {getLabelDate(this.props.todo.createdDate)}
                        </Label>

                        <Button size='small' floated='right' icon='delete' onClick={() => this.props.handleDelete(this.props.todo.id)} />
                    </List.Item>
                </div>
                <div>
                    <TodoTagList todoTags={tagArray} handleDeleteTodoTag={this.props.handleDeleteTodoTag} />
                </div>
            </div>
        );
    }
}

Todo.propTypes = {
    handleDelete: func.isRequired,
    handleDeleteTodoTag: func.isRequired,
    handleComplete: func.isRequired,
    sortByDateAndCompleted: func.isRequired,
    addTodoTag: func.isRequired,
    todo: shape({
        id: string.isRequired,
        title: string.isRequired,
        completed: bool.isRequired,
        createdDate: instanceOf(Date).isRequired,
        tags: arrayOf(shape({
            id: string.isRequired,
            title: string.isRequired,
            bindingId: number.isRequired}).isRequired).isRequired
    }).isRequired
};

export default Todo;