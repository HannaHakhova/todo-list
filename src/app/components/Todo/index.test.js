/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import {shallow} from 'enzyme';
import Todo from './index';

describe('Todo', () => {
    it('matches expected snapshot when completed is true)', () => {
        expect(shallow(<Todo
            handleDelete={() => {}}
            handleComplete={() => {}}
            handleDeleteTodoTag={() => {}}
            sortByDateAndCompleted={() => {}}
            addTodoTag={() => {}}
            todo={{id: '123456', title: 'test', completed: true, createdDate: new Date(1900, 1, 1)}} />)).toMatchSnapshot();
    });

    it('matches expected snapshot when completed is false) ', () => {
        expect(shallow(<Todo
            handleDelete={() => {}}
            handleComplete={() => {}}
            sortByDateAndCompleted={() => {}}
            addTodoTag={() => {}}
            handleDeleteTodoTag={() => {}}
            todo={{id: '123456', title: 'test', completed: false, createdDate: new Date(1900, 1, 1)}} />)).toMatchSnapshot();
    });

    it('calls handleComplete when checkbox is clicked', () => {
        const handleComplete = jest.fn();
        const wrapper = shallow(<Todo
            handleDelete={() => {}}
            handleComplete={handleComplete}
            sortByDateAndCompleted={() => {}}
            addTodoTag={() => {}}
            handleDeleteTodoTag={() => {}}
            todo={{id: '123456', title: 'test', completed: false, createdDate: new Date(1900, 1, 1)}} />);

        wrapper.find('Checkbox').simulate('change');

        expect(handleComplete).toBeCalledWith('123456');
    });

    it('calls sortByCreatedDateAndCompleted when label is clicked', () => {
        const sortByCreatedDateAndCompleted = jest.fn();
        const wrapper = shallow(<Todo
            handleDelete={() => {}}
            handleComplete={() => {}}
            sortByDateAndCompleted={sortByCreatedDateAndCompleted}
            addTodoTag={() => {}}
            handleDeleteTodoTag={() => {}}
            todo={{id: '123456', title: 'test', completed: false, createdDate: new Date(1900, 1, 1)}} />);

        wrapper.find('Label').simulate('click');

        expect(sortByCreatedDateAndCompleted).toBeCalled();
    });

    it('calls handleDelete when button is clicked', () => {
        const handleDelete = jest.fn();
        const wrapper = shallow(<Todo
            handleDelete={handleDelete}
            handleComplete={() => {}}
            sortByDateAndCompleted={() => {}}
            addTodoTag={() => {}}
            handleDeleteTodoTag={() => {}}
            todo={{id: '123456', title: 'test', completed: false, createdDate: new Date(1900, 1, 1)}} />);

        wrapper.find('Button').simulate('click');

        expect(handleDelete).toBeCalledWith('123456');
    });
});