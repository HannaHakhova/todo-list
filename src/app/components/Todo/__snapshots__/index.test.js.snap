// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo matches expected snapshot when completed is false)  1`] = `
<div
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <div>
    <ListItem
      floated="left"
    >
      <Checkbox
        checked={false}
        label="test"
        onChange={[Function]}
        type="checkbox"
      />
      <Label
        as="a"
        onClick={[Function]}
        size="tiny"
      >
        <Icon
          as="i"
          name="calendar"
        />
        2/1/1900
      </Label>
      <Button
        as="button"
        floated="right"
        icon="delete"
        onClick={[Function]}
        size="small"
      />
    </ListItem>
  </div>
  <div>
    <TodoTagList
      handleDeleteTodoTag={[Function]}
    />
  </div>
</div>
`;

exports[`Todo matches expected snapshot when completed is true) 1`] = `
<div
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <div>
    <ListItem
      floated="left"
    >
      <Checkbox
        checked={true}
        label="test"
        onChange={[Function]}
        type="checkbox"
      />
      <Label
        as="a"
        onClick={[Function]}
        size="tiny"
      >
        <Icon
          as="i"
          name="calendar"
        />
        2/1/1900
      </Label>
      <Button
        as="button"
        floated="right"
        icon="delete"
        onClick={[Function]}
        size="small"
      />
    </ListItem>
  </div>
  <div>
    <TodoTagList
      handleDeleteTodoTag={[Function]}
    />
  </div>
</div>
`;
